### DISCUSSION TITLE
Enable preview mode during development in an iframe

### DISCUSSION THREAD

**Discussion by dohomi** [vercel/next.js#32238] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.
Storyblok visual composer runs NextJS inside an IFrame. Due to this handling the cookie to enable preview mode is never set and the preview mode does not work for SSG handling.

## Describe the solution you'd like
Probably it would be a good idea if the preview mode could be enabled through URL params. The cookie solution might break other headless CMS systems as well

## Describe alternatives you've considered
Alternatively I'd like to run my site in SSR mode on a specific domain. It feels like a hack because I would need to deploy every project twice for each mode.

**Comment by markjoeljimenez** [comment]
+1! CraftCMS's preview mode is also using iFrames.

**Comment by timneutkens** [comment]
This is fixed by https://github.com/zeit/next.js/pull/11495 I guess?

**Comment by ChrisDalley** [comment]
From what I can see using Storyblok, this is still an issue @timneutkens after upgrading to latest canary build. 

The 307 response has the set-cookie headers with SameSite set to lax 
![image](https://user-images.githubusercontent.com/1242881/78149099-814ed900-742d-11ea-90c3-1601773baa3d.png)

But unlike running in localhost outside of Storyblok, the actual nextjs response doesn't contain the cookies in its request headers.
![image](https://user-images.githubusercontent.com/1242881/78149384-db4f9e80-742d-11ea-92ca-6451d8fa2e26.png)

**Comment by dohomi** [comment]
@timneutkens  on `NextJS 9.3.4` its still an issue in Storyblok, I have the exact behaviour as @ChrisDalley pointed out. Currently I use a SSR page and share the logic between SSG and SSR to solve the issue

**Comment by dohomi** [comment]
I tried with `9.3.6` the updated handling of the updated cms examples: 
```js
res.write(
    `<!DOCTYPE html><html><head><meta http-equiv="Refresh" content="0; url=${currentSlug}" />
    <script>window.location.href = '${currentSlug}'</script>
    </head>`
  )
res.end()
```
But the issue is still around, the previewData is getting lost. Did anybody got around this issue yet?

**Comment by ChrisDalley** [comment]
@dohomi - not yet, reverted back to using a separate server/routes for generating preview pages, though with Vercels new lambda # limits for hosting I would really like to get rid of them and onto the new preview functionality.

**Comment by dohomi** [comment]
I am using a SSR route now but on local environment this does not work at all: 
```
Error was not caught DOMException: Failed to read the 'localStorage' property from 'Window': Access is denied for this document.
```
This seems for me like a bug in the new version of NextJS (v. 9.4.x)

**Comment by peps** [comment]
Just a heads up, If you're using Brave, make sure to turn off Brave Shield on storyblok.com. It will block the cookies required for preview mode.

**Comment by rokinsky** [comment]
I'm having the same issue with Storyblok. @timneutkens, is it possible to add `sameSite` option in `res.setPreviewData`? 
[next/next-server/server/api-utils.ts#L467](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/api-utils.ts#L467)
[next/next-server/server/api-utils.ts#L476](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/api-utils.ts#L476)
It works fine when its property is empty (`none`), so it would be great to decide if someone wants to use `lax` by default.

**Comment by rokinsky** [comment]
> I'm having the same issue with Storyblok. @timneutkens, is it possible to add `sameSite` option in `res.setPreviewData`?
> [next/next-server/server/api-utils.ts#L467](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/api-utils.ts#L467)
> [next/next-server/server/api-utils.ts#L476](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/api-utils.ts#L476)
> It works fine when its property is empty (`none`), so it would be great to decide if someone wants to use `lax` by default.

btw here is a workaround:
```js
const IS_PROD = process.env.NODE_ENV === "production";

const setCookieSameSite = (res, value) => {
  const cookies = res.getHeader("Set-Cookie");
  res.setHeader(
    "Set-Cookie",
    cookies?.map((cookie) =>
      cookies.replace(
        cookie,
        "SameSite=Lax",
        `SameSite=${value}; ${IS_PROD ? "Secure;" : ""}`
      )
    )
  );
};

export const preview = async (req, res) => {
  // ...
  res.setPreviewData({});
  setCookieSameSite(res, "None");
  // ...
}
```
To use preview mode correctly on localhost, if you are using Chrome, disable Cookies without SameSite must be secure (chrome://flags/#cookies-without-same-site-must-be-secure) option.

**Comment by highflying** [comment]
> btw here is a workaround:
> 
> ```js
> const setCookieSameSite = (res, value) => {
>   const cookies = res.getHeader('Set-Cookie');
>   res.setHeader('Set-Cookie', cookies?.map((cookie) => cookie.replace("SameSite=Lax", `SameSite=${value}`)))
> }
> 
> export const preview = async (req, res) => {
>   // ...
>   res.setPreviewData({});
>   setCookieSameSite(res, "None");
>   // ...
> }
> ```

Unfortunately this doesn't work in the latest chrome, it complains that `secure` also needs to be set.

**Comment by rokinsky** [comment]
> > btw here is a workaround:
> > ```js
> > const setCookieSameSite = (res, value) => {
> >   const cookies = res.getHeader('Set-Cookie');
> >   res.setHeader('Set-Cookie', cookies?.map((cookie) => cookie.replace("SameSite=Lax", `SameSite=${value}`)))
> > }
> > 
> > export const preview = async (req, res) => {
> >   // ...
> >   res.setPreviewData({});
> >   setCookieSameSite(res, "None");
> >   // ...
> > }
> > ```
> 
> Unfortunately this doesn't work in the latest chrome, it complains that `secure` also needs to be set.

Thanks. I updated the comment.

**Comment by fredrik-sogaard** [comment]
Does anyone have a good workaround for this issue? We want to use Craft CMS in headless mode, but this issue makes the preview feature a bit cumbersome to work with.

**Comment by mutewinter** [comment]
The only workaround I found that works for Safari and other third party cookie blockers was to host my CMS on the same domain as the site. Obviously for some CMSes this isn’t possible.

**Comment by codercatdev** [comment]
Not familiar with when a issue moves to discussion, is this still a problem @dohomi for you?

I can’t figure out how to preview when using builder.io as it also iFrames in the site. Right now I have to set another flag after publishing to make it “published” but in “preview”

**Reply by dohomi** [reply]
@codercatdev the issue is not present any longer for Storyblok users.

**Reply by codercatdev** [reply]
@dohomi did Storyblok change, or was it fixed in Next.js somehow?

**Reply by dohomi** [reply]
@codercatdev check out https://github.com/vercel/next.js/discussions/32238#discussioncomment-1766968 that was handling the cookie token issue. It needs the correct cookie token handling in order to work

**Reply by codercatdev** [reply]
Thank you I missed the `Secure` part of this equation. I think we actually have Builder passing secure even on localhost and that was catching me up :D
